[Xcode] Stop passing build settings on the command line to avoid invalidating IDE builds

emw-apple opened the pull request:
#### 05696d76ec9ef470c91faa8b2c90101cae225d0e
<pre>
[Xcode] Stop passing build settings on the command line to avoid invalidating IDE builds
rdar://88135402

Reviewed by Jonathan Bedard and Alexey Proskuryakov.

Xcode considers a shell script&apos;s environment part of its task
signature, so changing build settings will require scripts to re-run in
the next build. Since many of our upstream build tasks (e.g. headers,
generated sources) are produced by script executions, changing build
settings is effectively a full build, so we must avoid it whenever
possible.

In support of this goal, update Make and build-webkit to stop passing
some build settings overrides by default. Build settings overrides which
can be expected to match what Xcode is using (e.g. SYMROOT, OBJROOT,
ARCHS, SDKROOT) are fine to keep passing.

* Makefile.shared: Only pass GCC_PREPROCESSOR_DEFINITIONS if there are
  actual custom arguments specified. Avoid passing
  WK_VALIDATE_DEPENDENCIES and COLOR_DIAGNOSTICS.
* Tools/CISupport/build-webkit-org/steps.py:
(CompileWebKit.start): As WK_VALIDATE_DEPENDENCIES is no longer supplied
  by default, enable it explicitly for bots. We should revisit this
  after workspace builds have been deployed and we can reasonably expect
  all Xcode-based builds of WebKit to occur via the workspace.
* Tools/CISupport/ews-build/steps.py:
(CompileWebKit.start): Ditto above.
* Tools/Scripts/build-webkit: Avoid passing WK_VALIDATE_DEPENDENCIES.

Canonical link: <a href="https://commits.webkit.org/252143@main">https://commits.webkit.org/252143@main</a>
</pre>

aproskuryakov commented on Makefile.shared:
```diff
@@ -9,7 +10,10 @@ ifeq ($(USE_WORKSPACE),YES)
 WORKSPACE_PATH ?= $(dir $(lastword $(MAKEFILE_LIST)))WebKit.xcworkspace
 XCODE_TARGET = -workspace $(WORKSPACE_PATH) -scheme "$(SCHEME)"
-XCODE_OPTIONS += WK_VALIDATE_DEPENDENCIES=YES_ERROR
+# FIXME: Move this setting to xcconfigs once all local Xcode builds of WebKit
+# happen in the workspace. When this is only passed on the command line, it
```
How far are we from this being the case?

emw-apple replied:
Quite close! Once Make uses workspace builds by default, engineers will need to go out of their way to build outside of a workspace.

emw-apple committed:
[Xcode] Stop passing build settings on the command line to avoid invalidating IDE builds
rdar://88135402

Reviewed by Jonathan Bedard and Alexey Proskuryakov.

Xcode considers a shell script's environment part of its task
signature, so changing build settings will require scripts to re-run in
the next build. Since many of our upstream build tasks (e.g. headers,
generated sources) are produced by script executions, changing build
settings is effectively a full build, so we must avoid it whenever
possible.

In support of this goal, update Make and build-webkit to stop passing
some build settings overrides by default. Build settings overrides which
can be expected to match what Xcode is using (e.g. SYMROOT, OBJROOT,
ARCHS, SDKROOT) are fine to keep passing.

* Makefile.shared: Only pass GCC_PREPROCESSOR_DEFINITIONS if there are
  actual custom arguments specified. Avoid passing
  WK_VALIDATE_DEPENDENCIES and COLOR_DIAGNOSTICS.
* Tools/CISupport/build-webkit-org/steps.py:
(CompileWebKit.start): As WK_VALIDATE_DEPENDENCIES is no longer supplied
  by default, enable it explicitly for bots. We should revisit this
  after workspace builds have been deployed and we can reasonably expect
  all Xcode-based builds of WebKit to occur via the workspace.
* Tools/CISupport/ews-build/steps.py:
(CompileWebKit.start): Ditto above.
* Tools/Scripts/build-webkit: Avoid passing WK_VALIDATE_DEPENDENCIES.

Canonical link: https://commits.webkit.org/252143@main
